### PR TITLE
Restructure library to conform to modern standards and best-practices

### PR DESCRIPTION
**Pull request by cbj4074**
This entire library should be rewritten, conforming to modern standards (e.g., PSRs) and best-practices, with accompanying tests.

This PR serves as a starting-point, and aims to "get us to a good place", so we're in a position to articulate the requirements through tests and then rework the code-base until they're all passing.

### CODE REVIEW

**Review by poppabear8883 — APPROVED** [review]

File: src/ProcessFailedException.php [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace IndieHD\AudioManipulator;
+
+use Symfony\Component\Process\Exception\ProcessFailedException as SymfonyProcessFailedException;
+
+class ProcessFailedException extends SymfonyProcessFailedException
+{
+    public function __construct(Process $process)
```
Comment (poppabear8883) [inline src/ProcessFailedException.php]: Might be a good idea to code to the `ProcessInterface` here ?

**Review by poppabear8883 — APPROVED** [review]

File: src/ProcessFailedException.php (hunk 1 above)
Comment (cbj4074) [reply src/ProcessFailedException.php]: Good catch; fixed in a subsequent commit.

File: src/Transcoding/Transcoder.php [hunk 2]
```diff
@@ -39,19 +43,19 @@ public function __construct(
...
-        $fadeInLength = 6,
-        $fadeOut = false,
-        $fadeOutLength = 6
+        string $inputFile,
+        string $outputFile,
+        float $trimStartTime = 0,
+        float $clipLength = 90,
+        float $fadeInLength = 0,
+        float $fadeOutLength = 0
     ) {
```
Comment (poppabear8883) [inline src/Transcoding/Transcoder.php]: If opting to use `parameter types`, maybe its note worthy to mention `return types` ?

Comment (cbj4074) [reply src/Transcoding/Transcoder.php]: Agreed. I've added them in a couple of places, and think it makes sense to use them consistently throughout. This is our chance to modernize the library, and there's no incentive not to take advantage of this construct.

File: tests/TranscodingTest.php [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    /**
+     * Ensure that a FLAC file can be transcoded to an MP3 file.
+     *
+     * @return void
+     */
+    public function testTranscodingFlacToMp3Succeeds()
+    {
+        $this->testDir =
```
Comment (poppabear8883) [inline tests/TranscodingTest.php]: what is this ??

Comment (cbj4074) [reply tests/TranscodingTest.php]: An oversight. Oops! It should simply be removed.

**Review by poppabear8883 — CHANGES_REQUESTED** [review]
Remove `Line #30` of `test/TranscodingTest.php`

**Review by poppabear8883 — CHANGES_REQUESTED** [review]
I would like to see a SoxCommand builder object before this gets merged in :wink:

File: src/CliCommand/SoxCommand.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        'effect' => [],     // Effect
+        'effopt' => [],     // Effect options
+    ];
+
+    public function name(): string
+    {
+        return $this->name;
+    }
+
+    public function addPart(string $name, string $value): void
```
Comment (poppabear8883) [inline src/CliCommand/SoxCommand.php]: So seeing your approach here I would recommend refactoring the method name from `addPart()` .. I'm not real sure what to name it but you're not really adding the "part" but rather attaching to a part in the parts array ?

File: src/CliCommand/SoxCommand.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+use IndieHD\AudioManipulator\CliCommand\SoxCommandInterface;
+
+class SoxCommand extends CliCommand implements SoxCommandInterface
+{
+    private $name = 'sox_command';
+
+    protected $binary = 'sox';
+
+    protected $parts = [
```
Comment (poppabear8883) [inline src/CliCommand/SoxCommand.php:13]: Make this an over-writable property in the base class

File: src/CliCommand/SoxCommand.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+namespace IndieHD\AudioManipulator\CliCommand;
+
+use IndieHD\AudioManipulator\CliCommand\SoxCommandInterface;
+
+class SoxCommand extends CliCommand implements SoxCommandInterface
+{
+    private $name = 'sox_command';
+
+    protected $binary = 'sox';
```
Comment (poppabear8883) [inline src/CliCommand/SoxCommand.php:11]: Make this an over-writable property in the base class

File: src/CliCommand/SoxCommand.php [hunk 7]
```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace IndieHD\AudioManipulator\CliCommand;
+
+use IndieHD\AudioManipulator\CliCommand\SoxCommandInterface;
+
+class SoxCommand extends CliCommand implements SoxCommandInterface
+{
+    private $name = 'sox_command';
```
Comment (poppabear8883) [inline src/CliCommand/SoxCommand.php]: Make this an over-writable property in the base class

File: src/CliCommand/SoxCommandInterface.php [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace IndieHD\AudioManipulator\CliCommand;
+
+use IndieHD\AudioManipulator\CliCommand\CliCommandInterface;
+
+interface SoxCommandInterface extends CliCommandInterface
+{
+
```
Comment (poppabear8883) [inline src/CliCommand/SoxCommandInterface.php:10]: Add your methods here :wink:

**Review by poppabear8883 — COMMENTED** [review]
Couple comments, Overall the approach seems to be on point with what was discussed, some fine tuning and testing should reveal what needs to be accounted for.

File: src/CliCommand/SoxCommand.php (hunk 4 above)
Comment (cbj4074) [reply src/CliCommand/SoxCommand.php]: Good call; you're absolutely correct. What about `addArgumentToPart()`, or even just `addArgument()`?

File: src/CliCommand/SoxCommand.php (hunk 5 above)
Comment (cbj4074) [reply src/CliCommand/SoxCommand.php:13]: What should its initial value be? An empty array, e.g., `protected $parts = [];`?

File: src/CliCommand/SoxCommand.php (hunk 6 above)
Comment (cbj4074) [reply src/CliCommand/SoxCommand.php:11]: Same question here; should the value be initialized to an empty string? `null`?

File: src/CliCommand/SoxCommand.php (hunk 4 above)
Comment (poppabear8883) [reply src/CliCommand/SoxCommand.php]: addArgument(); sounds good to me ?

File: src/CliCommand/SoxCommand.php (hunk 5 above)
Comment (poppabear8883) [reply src/CliCommand/SoxCommand.php:13]: make it an abstract property (forcing the implementation to overwrite it) and yes an empty array

Comment (poppabear8883) [reply src/CliCommand/SoxCommand.php:13]: ignore that 😕

File: src/CliCommand/SoxCommandInterface.php (hunk 8 above)
Comment (cbj4074) [reply src/CliCommand/SoxCommandInterface.php:10]: Oops! Thanks. :)

File: src/CliCommand/SoxCommand.php (hunk 5 above)
Comment (poppabear8883) [reply src/CliCommand/SoxCommand.php:13]: yes initialize it to an empty array or `null` ? if `null` then we can `test` if its been set in the implementation, if not through an exception

**Comment by cbj4074** [comment]
> 
> 
> I would like to see a SoxCommand builder object before this gets merged in 😉

I've implemented that. I'm not 100% happy with it yet (I'll explain why in a subsequent note), but it's a start.

**Review by poppabear8883 — APPROVED** [review]

**Pull request merged**